linux: Add definitions for Multipath TCP

george-hopkins opened the pull request:
The Linux kernel recently gained support for [Multipath TCP](https://en.wikipedia.org/wiki/Multipath_TCP), a protocol which builds upon TCP. This patch adds `IPPROTO_MPTCP` to open new sockets and `TCP_ULP` to query the upper layer protocol of traditional TCP sockets (which is set to `mptcp` for all underlying connections)

Source: [tcp: Define IPPROTO_MPTCP](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=faf391c3826cd29feae02078ca2022d2f912f7cc)

cc #1896

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @JohnTitor (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/libc/blob/master/CONTRIBUTING.md) for more information.

JohnTitor commented:
This is a _breaking change_ and I think we should add a deprecation note to tell users that the value will be changed here.

bors commented:
:umbrella: The latest upstream changes (presumably #1889) made this pull request unmergeable. Please resolve the merge conflicts.

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

george-hopkins commented:
Thank you for the review. I rebased the changes onto #1889 and added a deprecation notice. Is this what you had in mind or should the warning be placed more prominently?

george-hopkins commented:
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author

JohnTitor commented on src/unix/linux_like/linux/mod.rs:
```diff
@@ -1262,6 +1263,15 @@ pub const SCHED_RESET_ON_FORK: ::c_int = 0x40000000;
+    since = "0.2.79",
+    note = "IPPROTO_MAX depends on the kernel version on the target system"
+)]
+pub const IPPROTO_MAX: ::c_int = 263;
```
This change will affect the users who use 5.5 or lower kernel, right? If so, I'd avoid increasing it in this PR and just put a deprecation note like "this value will be changed following to the upstream change after a few releases.".

george-hopkins replied:
Yes, in these cases the value defined here won't match the actual value used by the kernel anymore. However, given that users might use `IPPROTO_MAX` to validate `IPPROTO_*` values, it think it would be preferable if the value would reflect the definition of `IPPROTO_MPTCP`.

JohnTitor replied:
So, anyway I think we should tell users this change before we go ahead. Generally, I wouldn't like to land any breaking changes without notice.

george-hopkins replied:
I understand that the change is less than ideal. For the time being, I set it back to its original value. If you'd like I will create a new pull request which will increase it again and can be merged as soon as other breaking changes are introduced.

george-hopkins commented:
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author

JohnTitor commented on src/unix/linux_like/linux/mod.rs:
```diff
@@ -1262,6 +1263,15 @@ pub const SCHED_RESET_ON_FORK: ::c_int = 0x40000000;
+/// Multipath TCP
+pub const IPPROTO_MPTCP: ::c_int = 262;
+#[deprecated(
+    since = "0.2.79",
```
```suggestion
    since = "0.2.80",
```

JohnTitor commented on src/unix/linux_like/linux/mod.rs:
```diff
@@ -1262,6 +1263,15 @@ pub const SCHED_RESET_ON_FORK: ::c_int = 0x40000000;
+    since = "0.2.79",
+    note = "IPPROTO_MAX depends on the kernel version on the target system"
+)]
+pub const IPPROTO_MAX: ::c_int = 263;
```
```suggestion
pub const IPPROTO_MAX: ::c_int = 256;
```

JohnTitor commented on src/unix/linux_like/linux/mod.rs:
```diff
@@ -1262,6 +1263,15 @@ pub const SCHED_RESET_ON_FORK: ::c_int = 0x40000000;
+pub const IPPROTO_MPTCP: ::c_int = 262;
+#[deprecated(
+    since = "0.2.79",
+    note = "IPPROTO_MAX depends on the kernel version on the target system"
```
```suggestion
    note = "This value was increased in the newer kernel \
            and we'll change this following upstream in the future release. \
            See #1896 for more info."
```

JohnTitor replied:
We should clarify we'll change this value in the future.

JohnTitor reviewed (commented):
r=me once all comments are addressed.

JohnTitor commented:
Thanks!
@bors r+

bors commented:
:pushpin: Commit 7f2619fdf93581a2822a3f63035e05f55d0df8be has been approved by `JohnTitor`

<!-- @bors r=JohnTitor 7f2619fdf93581a2822a3f63035e05f55d0df8be -->
<!-- homu: {"type":"Approved","sha":"7f2619fdf93581a2822a3f63035e05f55d0df8be","approver":"JohnTitor"} -->

bors commented:
:hourglass: Testing commit 7f2619fdf93581a2822a3f63035e05f55d0df8be with merge e0a4fb03978eceb295a91cea341e66b6163b56fd...
<!-- homu: {"type":"BuildStarted","head_sha":"7f2619fdf93581a2822a3f63035e05f55d0df8be","merge_sha":"e0a4fb03978eceb295a91cea341e66b6163b56fd"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/libc/runs/1303549497), [checks-cirrus-freebsd-11](https://cirrus-ci.com/task/5602568736866304), [checks-cirrus-freebsd-12](https://cirrus-ci.com/task/6728468643708928), [checks-cirrus-freebsd-13](https://cirrus-ci.com/task/4529445388156928)
Approved by: JohnTitor
Pushing e0a4fb03978eceb295a91cea341e66b6163b56fd to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"JohnTitor","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/libc/runs/1303549497","checks-cirrus-freebsd-11":"https://cirrus-ci.com/task/5602568736866304","checks-cirrus-freebsd-12":"https://cirrus-ci.com/task/6728468643708928","checks-cirrus-freebsd-13":"https://cirrus-ci.com/task/4529445388156928"},"merge_sha":"e0a4fb03978eceb295a91cea341e66b6163b56fd"} -->
